crates/dbsp/src/trace/spine_async/mod.rs

```diff
@@ -102,29 +102,26 @@ where
     B: Batch,
 {
     /// If this slot doesn't currently have an ongoing merge, and it does have
-    /// at least two loose batches, picks two of the loose batches and makes
+    /// at least two loose batches, picks an upper limit of the loose batches and makes
     /// them into merging batches, and returns those batches. Otherwise, returns
     /// `None` without changing anything.
     ///
-    /// We merge the least recently added batch (ensuring that batches
-    /// eventually get merged) with the one closest in size (ensuring that the
-    /// merge is as efficient as it can be).
-    fn try_start_merge(&mut self) -> Option<[Arc<B>; 2]> {
+    /// We merge the least recently added batches (ensuring that batches
+    /// eventually get merged).
+    fn try_start_merge(&mut self, level: usize) -> Option<Vec<Arc<B>>> {
+        // Lower level slots merge more batches at once. This is what other LSM implementations do
+        // although with slightly different numbers. The rationale is that the lower levels,
+        // which have more batches, should merge more batches at once. It's not clear why not every
+        // level should merge many batches.
+        const MAX_BATCHES_MERGED_AT_ONCE: [usize; MAX_LEVELS] = [12usize, 12, 12, 12, 12, 12, 12, 12, 12];
+
         if self.merging_batches.is_none() && self.loose_batches.len() >= 2 {
-            let a = self.loose_batches.pop_front().unwrap();
-            let a_len = a.len();
-            let mut best_idx = 0;
-            let mut best_distance = self.loose_batches[0].len().abs_diff(a_len);
-            for idx in 1..self.loose_batches.len() {
-                let distance = self.loose_batches[idx].len().abs_diff(a_len);
-                if distance < best_distance {
-                    best_idx = idx;
-                    best_distance = distance;
-                }
+            let mut batches = Vec::with_capacity(MAX_BATCHES_MERGED_AT_ONCE[level]);
+            for _i in 0..std::cmp::min(MAX_BATCHES_MERGED_AT_ONCE[level], self.loose_batches.len()) {
+                batches.push(self.loose_batches.pop_front().unwrap());
             }
```

Comment:
Apologies for nitpicking but this code bugged me, I would do something more like this:

```rust
            let n = std::cmp::min(MAX_BATCHES_MERGED_AT_ONCE[level], self.loose_batches.len());
            let batches = self.loose_batches.drain(..n).collect::<Vec<_>>();
```